File: map.c
#include <time.h>
#include <stdlib.h>
#include "map.h"

static char def_map_map_buffer[] = 
		"****************************"
		"*............**............*"
		"*.****.*****.**.*****.****.*"
		"*o****.*****.**.*****.****o*"
		"*.****.*****.**.*****.****.*"
		"*..........................*"
		"*.****.**.********.**.****.*"
		"*.****.**.********.**.****.*"
		"*......**....**....**......*"
		"******.***** ** *****.******"
		"     *.***** ** *****.*     "
		"     *.**          **.*     "
		"     *.** ***__*** **.*     "
		"******.** **    ** **.******"
		"      .   **    **   .      "
		"******.** ******** **.******"
		"     *.** ******** **.*     "
		"     *.**          **.*     "
		"     *.** ******** **.*     "
		"******.** ******** **.******"
		"*............**............*"
		"*.****.*****.**.*****.****.*"
		"*.****.*****.**.*****.****.*"
		"*o..**................**..o*"
		"***.**.**.********.**.**.***"
		"***.**.**.********.**.**.***"
		"*......**....**....**......*"
		"*.**********.**.**********.*"
		"*.**********.**.**********.*"
		"*..........................*"
		"****************************";

static const map_t def_map = {
	.width = 28,
	.height = 31,
	.cell_width = 0,
	.cell_height = 0,
	.food_count = 246,
	.buffer = 0
};

static int at(map_t *m, int x, int y)
{
	if(x >= m->width || x < 0) return NONE;
	if(y >= m->height || y < 0) return NONE;

	return m->buffer[y * m->width + x];
}

int(*map_pos_at)(map_t *m, int x, int y);

map_t map_create_default(int width, int height)
{
	srand(time(NULL));
	map_t m = def_map;
	m.cell_width = width / m.width;
	m.cell_height = height / m.height;

	map_pos_at = at;

	if(m.buffer == NULL)
		m.buffer = calloc(sizeof(char), m.width * m.height);
	memcpy(m.buffer, def_map_map_buffer, sizeof def_map_map_buffer);

	int x = 0, y = 0;
	for(int i = 0; i < 6; i++)
	{
		x = (rand() % m.width);
		y = (rand() % m.height);

		if(m.buffer[y * m.width + x] == NORMAL_FOOD)
			m.buffer[y * m.width + x] = BIG_FOOD;
	}

	return m;
}

void map_draw(map_t *map, SDL_Renderer *render, uint32_t color)
{
	const int lw = map->cell_width;
	const int lh = map->cell_height;

	const int ldw = lw * 10 / 29;
	const int ldh = lh * 10 / 29;

	// SDL_Rect r;

	for(int x = 0; x < map->width; x++)
	{
		for(int y = 0; y < map->height; y++)
		{
			const int lx = x * lw;
			const int ly = y * lh;

			// r.x = lx; r.y = ly;
			// r.w = lw; r.h = lh;

			switch(at(map, x, y))
			{
				case NONE:
					break;


				case WALL:
					{
						SDL_SetRenderDrawColor(render, (color >> 16) & 0xFF, (color >> 8) & 0xFF, color & 0xFF, 0);

						int c = 1;

						if(at(map, x + 1, y) != WALL)
						{
							SDL_RenderDrawLine(render,
									lx + lw - ldw,
									ly + ldh * (at(map, x, y - 1) != WALL) ,
									lx + lw - ldw,
									ly + lh - ldh * (at(map, x, y + 1) != WALL));
						}
						else
							c *= 2;

						if(at(map, x - 1, y) != WALL)
						{
							SDL_RenderDrawLine(render,
									lx + ldw,
									ly + ldh * (at(map, x, y - 1) != WALL),
									lx + ldw,
									ly + lh - ldh * (at(map, x, y + 1) != WALL));
						}
						else
							c *= 3;

						if(at(map, x, y + 1) != WALL)
						{
							SDL_RenderDrawLine(render,
									lx + ldw * (at(map, x - 1, y) != WALL),
									ly + lh - ldh,
									lx + lw - ldw * (at(map, x + 1, y) != WALL),
									ly + lh - ldh);
						}
						else
							c *= 5;

						if(at(map, x, y - 1) != WALL)
						{
							SDL_RenderDrawLine(render,
									lx + ldw * (at(map, x - 1, y) != WALL),
									ly + ldh,
									lx + lw - ldw * (at(map, x + 1, y) != WALL),
									ly + ldh);
							
						}
						else
							c *= 7;

						if(c % 21 == 0) SDL_RenderDrawLine(render, lx + ldw, ly, lx, ly + ldh);
						if(c % 14 == 0) SDL_RenderDrawLine(render, lx + lw - ldw, ly, lx + lw, ly + ldh);
						if(c % 10 == 0) SDL_RenderDrawLine(render, lx + lw - ldw, ly + lh, lx + lw, ly + lh - ldh);
						if(c % 15 == 0) SDL_RenderDrawLine(render, lx, ly + lh - ldh, lx + ldw, ly + lh);
					}
					break;

				case DOOR:
					SDL_SetRenderDrawColor(render, 0xff, 0xee, 0xee, 0);
					SDL_RenderDrawLine(render, lx, ly + ldh, lx + lw, ly + ldh);
					break;

				case NORMAL_FOOD:
					{
						SDL_SetRenderDrawColor(render, 0xff, 0xff, 0xae, 0);
						SDL_Rect food = {lx + 5 * lw/12, ly + 5 * lh/12, lw/6, lh/6};
						SDL_RenderFillRect(render, &food);
					}
					break;

				case BIG_FOOD:
					{
						SDL_SetRenderDrawColor(render, 0xff, 0xff, 0xae, 0);
						SDL_Rect food = {lx + lw/3, ly + lh/3, lw/3, lh/3};
						SDL_RenderFillRect(render, &food);
					}
					break;
			}
		}
	}
}

File: map.js
const NORMAL_FOOD = 9
const BIG_FOOD = 8
const NONE = 7
const DOOR = 6
const WALL = 5

const default_map = 
[
5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 
5, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 5, 5, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 5, 
5, 9, 5, 5, 5, 5, 9, 5, 5, 5, 5, 5, 9, 5, 5, 9, 5, 5, 5, 5, 5, 9, 5, 5, 5, 5, 9, 5, 
5, 8, 5, 5, 5, 5, 9, 5, 5, 5, 5, 5, 9, 5, 5, 9, 5, 5, 5, 5, 5, 9, 5, 5, 5, 5, 8, 5, 
5, 9, 5, 5, 5, 5, 9, 5, 5, 5, 5, 5, 9, 5, 5, 9, 5, 5, 5, 5, 5, 9, 5, 5, 5, 5, 9, 5, 
5, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 5, 
5, 9, 5, 5, 5, 5, 9, 5, 5, 9, 5, 5, 5, 5, 5, 5, 5, 5, 9, 5, 5, 9, 5, 5, 5, 5, 9, 5, 
5, 9, 5, 5, 5, 5, 9, 5, 5, 9, 5, 5, 5, 5, 5, 5, 5, 5, 9, 5, 5, 9, 5, 5, 5, 5, 9, 5, 
5, 9, 9, 9, 9, 9, 9, 5, 5, 9, 9, 9, 9, 5, 5, 9, 9, 9, 9, 5, 5, 9, 9, 9, 9, 9, 9, 5, 
5, 5, 5, 5, 5, 5, 9, 5, 5, 5, 5, 5, 7, 5, 5, 7, 5, 5, 5, 5, 5, 9, 5, 5, 5, 5, 5, 5, 
7, 7, 7, 7, 7, 5, 9, 5, 5, 5, 5, 5, 7, 5, 5, 7, 5, 5, 5, 5, 5, 9, 5, 7, 7, 7, 7, 7, 
7, 7, 7, 7, 7, 5, 9, 5, 5, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 5, 5, 9, 5, 7, 7, 7, 7, 7, 
7, 7, 7, 7, 7, 5, 9, 5, 5, 7, 5, 5, 5, 6, 6, 5, 5, 5, 7, 5, 5, 9, 5, 7, 7, 7, 7, 7, 
5, 5, 5, 5, 5, 5, 9, 5, 5, 7, 5, 5, 7, 7, 7, 7, 5, 5, 7, 5, 5, 9, 5, 5, 5, 5, 5, 5, 
7, 7, 7, 7, 7, 7, 9, 7, 7, 7, 5, 5, 7, 7, 7, 7, 5, 5, 7, 7, 7, 9, 7, 7, 7, 7, 7, 7, 
5, 5, 5, 5, 5, 5, 9, 5, 5, 7, 5, 5, 5, 5, 5, 5, 5, 5, 7, 5, 5, 9, 5, 5, 5, 5, 5, 5, 
7, 7, 7, 7, 7, 5, 9, 5, 5, 7, 5, 5, 5, 5, 5, 5, 5, 5, 7, 5, 5, 9, 5, 7, 7, 7, 7, 7, 
7, 7, 7, 7, 7, 5, 9, 5, 5, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 5, 5, 9, 5, 7, 7, 7, 7, 7, 
7, 7, 7, 7, 7, 5, 9, 5, 5, 7, 5, 5, 5, 5, 5, 5, 5, 5, 7, 5, 5, 9, 5, 7, 7, 7, 7, 7, 
5, 5, 5, 5, 5, 5, 9, 5, 5, 7, 5, 5, 5, 5, 5, 5, 5, 5, 7, 5, 5, 9, 5, 5, 5, 5, 5, 5, 
5, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 5, 5, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 5, 
5, 9, 5, 5, 5, 5, 9, 5, 5, 5, 5, 5, 9, 5, 5, 9, 5, 5, 5, 5, 5, 9, 5, 5, 5, 5, 9, 5, 
5, 9, 5, 5, 5, 5, 9, 5, 5, 5, 5, 5, 9, 5, 5, 9, 5, 5, 5, 5, 5, 9, 5, 5, 5, 5, 9, 5, 
5, 8, 9, 9, 5, 5, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 5, 5, 9, 9, 8, 5, 
5, 5, 5, 9, 5, 5, 9, 5, 5, 9, 5, 5, 5, 5, 5, 5, 5, 5, 9, 5, 5, 9, 5, 5, 9, 5, 5, 5, 
5, 5, 5, 9, 5, 5, 9, 5, 5, 9, 5, 5, 5, 5, 5, 5, 5, 5, 9, 5, 5, 9, 5, 5, 9, 5, 5, 5, 
5, 9, 9, 9, 9, 9, 9, 5, 5, 9, 9, 9, 9, 5, 5, 9, 9, 9, 9, 5, 5, 9, 9, 9, 9, 9, 9, 5, 
5, 9, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 9, 5, 5, 9, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 9, 5, 
5, 9, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 9, 5, 5, 9, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 9, 5, 
5, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 5, 
5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5]

const colors = ['magenta', 'lime', 'red', 'yellow']

class Map
{
	constructor(width, height)
	{
		this.width = 28
		this.height = 31
		this.cell_width = width / this.width
		this.cell_height = height / this.height
		this.food_count = 246
		this.buffer = default_map.slice()

		for(let i = 0; i < 6; i++)
		{
			let x = parseInt(Math.random() * (this.width - 1))
			let y = parseInt(Math.random() * (this.height - 1))

			if(this.buffer[y * this.width + x] == NORMAL_FOOD)
				this.buffer[y * this.width + x] = BIG_FOOD
		}
	}

	at(x, y)
	{
		x = parseInt(x)
		y = parseInt(y)

		if(x >= this.width || x < 0) return NONE
		if(y >= this.height || y < 0) return NONE

		return this.buffer[y * this.width + x]
	}

	pat(x, y, value = NONE)
	{
		x = parseInt(x)
		y = parseInt(y)

		if(x >= this.width || x < 0) return
		if(y >= this.height || y < 0) return

		this.buffer[y * this.width + x] = value
	}

	draw(ctx, color = 'blue')
	{
		const lw = this.cell_width
		const lh = this.cell_height

		const ldw = lw * 10 / 29
		const ldh = lh * 10 / 29

		for(let x = 0; x < this.width; x++)
		{
			for(let y = 0; y < this.height; y++)
			{
				const lx = x * lw;
				const ly = y * lh;

				switch(this.at(x, y))
				{
					case NONE:
						break;

					case WALL:
						{
							ctx.beginPath()
							ctx.lineWidth = 2
							ctx.strokeStyle = color

							let c = 1;

							if(this.at(x + 1, y) != WALL)
							{
								ctx.moveTo(lx + lw - ldw, ly + (this.at(x, y - 1) != WALL ? ldh : 0))
								ctx.lineTo(lx + lw - ldw, ly + lh - (this.at(x, y + 1) != WALL ? ldh : 0))
							}
							else
								c *= 2;

							if(this.at(x - 1, y) != WALL)
							{
								ctx.moveTo(lx + ldw, ly + (this.at(x, y - 1) != WALL ? ldh : 0))
								ctx.lineTo(lx + ldw, ly + lh - (this.at(x, y + 1) != WALL ? ldh : 0))
							}
							else
								c *= 3;

							if(this.at(x, y + 1) != WALL)
							{
								ctx.moveTo(lx + (this.at(x - 1, y) != WALL ? ldw : 0), ly + lh - ldh)
								ctx.lineTo(lx + lw - (this.at(x + 1, y) != WALL ? ldw : 0), ly + lh - ldh)
							}
							else
								c *= 5;

							if(this.at(x, y - 1) != WALL)
							{
								ctx.moveTo(lx + (this.at(x - 1, y) != WALL ? ldw : 0), ly + ldh)
								ctx.lineTo(lx + lw - (this.at(x + 1, y) != WALL ? ldw : 0), ly + ldh)
							}
							else
								c *= 7;

							if(c % 21 == 0)
							{
								ctx.moveTo(lx + ldw, ly)
								ctx.lineTo(lx, ly + ldh)
							}
							if(c % 14 == 0)
							{
								ctx.moveTo(lx + lw - ldw, ly)
								ctx.lineTo(lx + lw, ly + ldh)
							}
							if(c % 10 == 0)
							{
								ctx.moveTo(lx + lw - ldw, ly + lh)
								ctx.lineTo( lx + lw, ly + lh - ldh)
							}
							if(c % 15 == 0)
							{
								ctx.moveTo(lx, ly + lh - ldh)
								ctx.lineTo(lx + ldw, ly + lh)
							}

							ctx.stroke();
						}
						break;

					case DOOR:
						{
							ctx.beginPath()
							ctx.lineWidth = 2
							ctx.strokeStyle = '#ffeeee'

							ctx.moveTo(lx, ly + ldh)
							ctx.lineTo(lx + lw, ly + ldh)

							ctx.stroke();
						}
						break;

					case NORMAL_FOOD:
						{
							ctx.fillStyle = '#ffffae'
							ctx.fillRect(lx + 5 * lw/12, ly + 5 * lh/12, lw/6, lh/6);
						}
						break;

					case BIG_FOOD:
						{
							ctx.fillStyle = colors[parseInt(performance.now() / 100) % color.length]
							ctx.fillRect(lx + lw * 0.3, ly + lh * 0.3, lw * 0.4, lh * 0.4);
						}
						break;
				}
			}
		}
	}
}
